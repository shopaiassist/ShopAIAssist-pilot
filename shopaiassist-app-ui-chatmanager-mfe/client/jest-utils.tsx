import React, { JSXElementConstructor, ReactElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouterProps } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';
import { merge } from 'ts-deepmerge';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { ByRoleMatcher, render, RenderResult } from '@testing-library/react';
import { reducer, RootState, store as baseStore } from './src/store/store';
import i18nInstance from './src/utils/i18n';

type InitialEntries = MemoryRouterProps['initialEntries'];
type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};
type PartialRootState = DeepPartial<RootState>;

interface RenderOptions {
  entries?: InitialEntries;
  state?: PartialRootState;
  store?: EnhancedStore;
}

interface SnapshotOptions {
  isModal?: boolean | ByRoleMatcher;
  state?: PartialRootState;
}

interface WrapperProps {
  children: ReactElement;
}

// export const mockApi = (apiInstance: AxiosInstance) => ({
//   delete: jest.spyOn(apiInstance, 'delete'),
//   get: jest.spyOn(apiInstance, 'get'),
//   post: jest.spyOn(apiInstance, 'post'),
//   put: jest.spyOn(apiInstance, 'put'),
//   patch: jest.spyOn(apiInstance, 'patch'),
// });

/** Redux store with optional default content */
export const mockStore = (state: PartialRootState = {}): EnhancedStore =>
  configureStore({ preloadedState: merge(baseStore.getState(), state), reducer: reducer as never });

/** Render function with access to override store/state */
export const renderAll = (ui: ReactElement, { entries, state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const Wrapper = ({ children }: WrapperProps) => (
    <MemoryRouter initialEntries={entries}>
      <Provider store={data}>
        <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
      </Provider>
    </MemoryRouter>
  );
  return render(ui, { wrapper: Wrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with i18n provider */
export const renderWithI18n = (ui: ReactElement): RenderResult => {
  const I18nWrapper = ({ children }: WrapperProps) => <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>;
  return render(ui, { wrapper: I18nWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with access to override store/state */
export const renderWithRedux = (ui: ReactElement, { state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const ReduxWrapper = ({ children }: WrapperProps) => (
    <Provider store={data}>
      <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
    </Provider>
  );
  return render(ui, { wrapper: ReduxWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Get content to test against a snapshot */
export const getSnapshot = (ui: ReactElement, state?: PartialRootState): ChildNode =>
  renderAll(ui, { state }).container.firstChild as ChildNode;

/** Get modal content to test against a snapshot */
export const getSnapshotModal = (ui: ReactElement, state?: PartialRootState, role?: ByRoleMatcher): ChildNode =>
  renderAll(ui, { state }).getByRole(role || 'presentation');

/** Helper function to generate snapshots */
export const snapshot = (ui: ReactElement, { isModal, state }: SnapshotOptions = {}) => {
  const role = typeof isModal === 'boolean' ? undefined : isModal;
  // eslint-disable-next-line no-extra-boolean-cast
  const snapshot = !!isModal ? getSnapshotModal(ui, state, role) : getSnapshot(ui, state);
  expect(snapshot).toMatchSnapshot();
};

// re-export everything
export * from '@testing-library/react';
