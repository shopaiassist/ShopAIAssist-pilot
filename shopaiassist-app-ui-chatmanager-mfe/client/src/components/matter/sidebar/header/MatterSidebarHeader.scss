@import "../../../../utils/styles/imports";

.matter-sidebar-header {
   .new-folder-or-chat-button-container {
      position: relative;

      .new-folder-or-chat-button {
         display: flex;
         gap: 0.5rem;
      }

      .new-folder-or-chat-menu {
         display: flex;
         flex-direction: column;
         margin-top: 0.5rem;
         z-index: map-get($z-index-map, 'default');

         .new-folder-or-chat-menu-item {
            display: flex;
            flex-direction: row;
            gap: 0.5rem;
            align-items: baseline;
         }
      }
   }
}
