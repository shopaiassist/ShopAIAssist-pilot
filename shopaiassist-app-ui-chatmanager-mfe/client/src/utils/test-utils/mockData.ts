import { TreeItem, FolderItem } from '../../@types/sidebar';

export const mockTree: (TreeItem | FolderItem)[] = [
  {
    treeItemId: '9a4144e0-65ea-4f6c-8cd5-43a0a6d1e634',
    name: 'Test Folder For Created At 2',
    type: 'folder',
    createdAt: new Date('2024-05-06T15:26:57.992Z'),
    updatedAt: new Date('2024-05-06T15:26:57.992Z'),
    fileCollectionId: 'whatever'
  },
  {
    treeItemId: '5aa13d2b-3f04-4a91-a84b-bf214420b47c',
    name: 'Test Folder For Created At 2',
    type: 'folder',
    createdAt: new Date('2024-05-06T15:19:07.430Z'),
    updatedAt: new Date('2024-05-06T15:19:07.430Z'),
    fileCollectionId: 'whatever'
  },
  {
    type: 'folder',
    name: 'New Folder 100',
    description: 'This folder is used for whatever, its created via postman.',
    matterId: 'Whatever Folder 100',
    treeItemId: '4c484fb3-faf6-4fe8-9930-410772f141d7',
    createdAt: new Date('2024-04-25T19:21:07.787Z'),
    updatedAt: new Date('2024-04-25T19:21:07.787Z'),
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-03-01T17:00:33.564Z'),
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b96',
    name: 'A chat named desire',
    type: 'chat',
    updatedAt: new Date('2024-05-06T18:09:49.245Z'),
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:30:33.080Z'),
    updatedAt: new Date('2024-02-23T16:30:33.080Z'),
    treeItemId: '8c1b78b3-a798-4557-871a-bff6c8ac804e',
    name: 'new2',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:30:17.078Z'),
    updatedAt: new Date('2024-02-23T16:30:17.078Z'),
    treeItemId: 'b39f37a3-6781-4445-8573-77658d9ee7ae',
    name: 'new1',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:26:17.254Z'),
    updatedAt: new Date('2024-02-23T16:26:17.254Z'),
    treeItemId: '9bc3c583-a6e4-41cc-92f1-27e1be3ba4f6',
    name: 'newnewenewnew',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-23T16:25:50.258Z'),
    updatedAt: new Date('2024-02-23T16:25:50.258Z'),
    treeItemId: '2ec6bba4-6dd0-4190-a015-b58e6ee4021c',
    name: '12346123445',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-21T20:51:41.137Z'),
    updatedAt: new Date('2024-02-21T20:51:41.137Z'),
    treeItemId: 'e4a9bc6a-cfb1-48ac-bc6e-1b6f135147c0',
    name: 'Chat Name 2',
    type: 'chat',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-20T19:47:03.738Z'),
    updatedAt: new Date('2024-02-20T19:47:03.738Z'),
    treeItemId: 'ce2422ea-b573-4c71-b569-fa2c79a5e50b',
    name: 'B matter with B name',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-20T19:44:31.642Z'),
    updatedAt: new Date('2024-02-20T19:44:31.642Z'),
    treeItemId: '148f72ac-b903-451c-88f2-ed1282c6acb1',
    name: 'Zmatter with Z name',
    matterId: 'A matter with A name',
    type: 'folder',
    fileCollectionId: 'whatever'
  },
  {
    createdAt: new Date('2024-02-15T18:35:41.269Z'),
    updatedAt: new Date('2024-02-15T18:35:41.269Z'),
    treeItemId: 'e82e81e6-68c8-4bd1-b67c-316dc4105d18',
    name: '123',
    matterId: 'Matter with a really REALLY long name, im mean like REALLY long',
    type: 'folder',
    fileCollectionId: 'whatever'
  }
];
