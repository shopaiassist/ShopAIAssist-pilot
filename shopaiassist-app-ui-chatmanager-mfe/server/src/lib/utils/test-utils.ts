import express, { NextFunction } from 'express';
import { AppRequest } from 'server/src/@types/appRequest';

/**
 * Sample data representing chat items for testing sorting and other functionalities.
 */
export const testChatsData = [
  {
    type: 'chat',
    name: 'New name for this chat',
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b97',
    createdAt: '2024-04-25T18:23:45.704Z',
    updatedAt: '2024-04-25T18:23:45.704Z',
    uid: '8beab79b628b479cb9c40855777a511d',
    parentId: '8c1b78b3-a798-4557-871a-bff6c8ac804e'
  },
  {
    createdAt: '2024-03-01T17:00:33.564Z',
    updatedAt: '2024-03-01T17:00:33.564Z',
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b96',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 1',
    type: 'chat'
  },
  {
    createdAt: '2024-02-21T20:51:41.137Z',
    updatedAt: '2024-02-21T20:51:41.137Z',
    treeItemId: 'e4a9bc6a-cfb1-48ac-bc6e-1b6f135147c0',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 2',
    type: 'chat'
  },
  {
    createdAt: '2024-02-21T20:49:41.137Z',
    updatedAt: '2024-02-21T20:49:41.137Z',
    treeItemId: 'c807ff14-58da-4199-8832-da90252fa743',
    parentId: 'b5014a24-8f61-42eb-a24e-8f0b88fd18ab',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Archived Chat',
    type: 'chat'
  }
];

/**
 * Sample data representing folder items for testing sorting and other functionalities.
 */
export const testFoldersData = [
  {
    createdAt: '2024-02-23T16:30:33.080Z',
    updatedAt: '2024-02-23T16:30:33.080Z',
    treeItemId: '8c1b78b3-a798-4557-871a-bff6c8ac804e',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new2',
    type: 'folder'
  },
  {
    createdAt: '2024-02-23T16:30:17.078Z',
    updatedAt: '2024-02-23T16:30:17.078Z',
    treeItemId: 'b39f37a3-6781-4445-8573-77658d9ee7ae',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new1',
    type: 'folder'
  },
  {
    createdAt: '2024-02-23T16:26:17.254Z',
    updatedAt: '2024-02-23T16:26:17.254Z',
    treeItemId: '9bc3c583-a6e4-41cc-92f1-27e1be3ba4f6',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'newnewenewnew',
    type: 'folder',
    parentId: '8c1b78b3-a798-4557-871a-bff6c8ac804e'
  },
  {
    createdAt: '2024-02-23T16:20:00.254Z',
    updatedAt: '2024-02-23T16:20:00.254Z',
    treeItemId: 'b5014a24-8f61-42eb-a24e-8f0b88fd18ab',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'archived folder',
    type: 'folder',
    isArchived: true
  }
];

/**
 * Sample data for testing sorting chat and folder items by date.
 */
export const chatsAndFoldersByDate = [
  {
    createdAt: '2024-03-01T17:00:33.564Z',
    updatedAt: '2024-03-01T17:00:33.564Z',
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b96',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 1',
    type: 'chat'
  },
  {
    createdAt: '2024-02-23T16:30:33.080Z',
    updatedAt: '2024-02-23T16:30:33.080Z',
    treeItemId: '8c1b78b3-a798-4557-871a-bff6c8ac804e',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new2',
    type: 'folder'
  },
  {
    createdAt: '2024-02-23T16:30:17.078Z',
    updatedAt: '2024-02-23T16:30:17.078Z',
    treeItemId: 'b39f37a3-6781-4445-8573-77658d9ee7ae',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new1',
    type: 'folder'
  },
  {
    createdAt: '2024-02-21T20:51:41.137Z',
    updatedAt: '2024-02-21T20:51:41.137Z',
    treeItemId: 'e4a9bc6a-cfb1-48ac-bc6e-1b6f135147c0',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 2',
    type: 'chat'
  }
];

/**
 * Sample data for testing sorting chat and folder items by name.
 */
export const chatsAndFoldersByName = [
  {
    createdAt: '2024-03-01T17:00:33.564Z',
    updatedAt: '2024-03-01T17:00:33.564Z',
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b96',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 1',
    type: 'chat'
  },
  {
    createdAt: '2024-02-21T20:51:41.137Z',
    updatedAt: '2024-02-21T20:51:41.137Z',
    treeItemId: 'e4a9bc6a-cfb1-48ac-bc6e-1b6f135147c0',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 2',
    type: 'chat'
  },
  {
    createdAt: '2024-02-23T16:30:17.078Z',
    updatedAt: '2024-02-23T16:30:17.078Z',
    treeItemId: 'b39f37a3-6781-4445-8573-77658d9ee7ae',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new1',
    type: 'folder'
  },
  {
    createdAt: '2024-02-23T16:30:33.080Z',
    updatedAt: '2024-02-23T16:30:33.080Z',
    treeItemId: '8c1b78b3-a798-4557-871a-bff6c8ac804e',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new2',
    type: 'folder'
  }
];

/**
 * Sample data for testing sorting chat and folder items by type.
 */
export const chatsAndFoldersByType = [
  {
    createdAt: '2024-03-01T17:00:33.564Z',
    updatedAt: '2024-03-01T17:00:33.564Z',
    treeItemId: 'e35607c7-be8f-4172-a3cd-ae7e39bc2b96',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 1',
    type: 'chat'
  },
  {
    createdAt: '2024-02-21T20:51:41.137Z',
    updatedAt: '2024-02-21T20:51:41.137Z',
    treeItemId: 'e4a9bc6a-cfb1-48ac-bc6e-1b6f135147c0',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Chat Name 2',
    type: 'chat'
  },
  {
    createdAt: '2024-02-23T16:30:33.080Z',
    updatedAt: '2024-02-23T16:30:33.080Z',
    treeItemId: '8c1b78b3-a798-4557-871a-bff6c8ac804e',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new2',
    type: 'folder'
  },
  {
    createdAt: '2024-02-23T16:30:17.078Z',
    updatedAt: '2024-02-23T16:30:17.078Z',
    treeItemId: 'b39f37a3-6781-4445-8573-77658d9ee7ae',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'new1',
    type: 'folder'
  }
];

/**
 * Sample data for testing sorting chat and folder items by type.
 */
export const archivedChatsAndFoldersByType = [
  {
    createdAt: '2024-02-23T16:20:00.254Z',
    updatedAt: '2024-02-23T16:20:00.254Z',
    treeItemId: 'b5014a24-8f61-42eb-a24e-8f0b88fd18ab',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'archived folder',
    type: 'folder',
    isArchived: true
  }
];

/**
 * Sample data for testing sorting chat and folder items by type.
 */
export const archivedSubChatsAndFoldersByType = [
  {
    createdAt: '2024-02-21T20:49:41.137Z',
    updatedAt: '2024-02-21T20:49:41.137Z',
    treeItemId: 'c807ff14-58da-4199-8832-da90252fa743',
    parentId: 'b5014a24-8f61-42eb-a24e-8f0b88fd18ab',
    uid: '8beab79b628b479cb9c40855777a511d',
    name: 'Archived Chat',
    type: 'chat'
  }
];

export const requestHeaders = [
  'authorization',
  'op-orch 0702241037270OsnGa3dYt2HQ63uQhF3xaVG9i3ctHLqykE6E8KI1ydoGl05xQPMYPzpz44SBTWG1nxQryAKW7OmWwsZUK9z31yAWY3Q91b1puRZso9CW5n7MF9HXhnuwqNOaeODJw5_bTwg_Ky2ZdAhEI6p9GKSSBirRcRxU4veo-beuiyJ7OPXaRqxy5uV2kahg9dp4_P2PYeIbVcQV-p9wGy2mlekRfCNIp8aY97kjQOGKuQdvOkbIHm3d8b_IgT90rf1QQ3BoOSaH-gX5odIDYxz8eJAl83wgn98x3RxJuYkowEQ_VzyWFM4x4fCHu7E_dgML2Xr1UpR4I-PaY5JL6DPGkQr3QA'
];

const ONEPASS_ORCHESTRATION_TOKEN_TYPE = 'op-orch';
const UNAUTHORIZED_BODY = { message: 'No authorized user', code: 'no_authorized_user' };
const INVALID_TOKEN_BODY = { message: 'Invalid authorization token type', code: 'invalid_auth_token_type' };

/**
 * This is a mock of the security middleware from Hades for use in unit tests.
 */
export class MockSecurityMiddleware {
  constructor() {}

  /**
   * Requires the user to be authenticated and adds the logged in user profile
   * and orchestration token to the express request object
   *
   * @param req   An express request with the `user` and `orchestrationToken` property added to it
   * @param res   An express response
   * @param next  An express next function
   */
  public isAuthenticated = async (req: AppRequest, res: express.Response, next: NextFunction) => {
    const userMap = new Map();
    userMap.set(requestHeaders[1], {
      email: 'hunter.nenneau+localtest1@thomsonreuters.com',
      firstName: 'Hunter',
      lastName: 'Nenneau',
      id: '8beab79b628b479cb9c40855777a511d',
      username: 'hunter.nenneau+localtest1@thomsonreuters.com',
      registrationKey: ''
    });
    try {
      const { authorization } = req.headers;
      if (authorization) {
        const [type, token] = authorization.split(' ');
        if (type === ONEPASS_ORCHESTRATION_TOKEN_TYPE) {
          try {
            req.orchestrationToken = token;
            req.user = userMap.get(authorization);
            next();
          } catch (err: unknown) {
            next(err);
          }
        } else if (type) {
          res.status(400).send(INVALID_TOKEN_BODY);
        }
      } else {
        res.status(401).send(UNAUTHORIZED_BODY);
      }
    } catch (err) {
      next(err);
    }
  };
}
