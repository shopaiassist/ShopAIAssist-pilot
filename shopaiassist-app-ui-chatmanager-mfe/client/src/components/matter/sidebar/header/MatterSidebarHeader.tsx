import { RefObject, useState } from 'react';
import { Button, Icon, MenuItem } from '@on/core-components/react';
import Translate from '../../../i18n/Translate';
import classnames from 'classnames';
import { useOutsideClick } from '../../../../hooks/useOutsideClick';
import { useSortMenu } from '../../../../hooks/useSortMenu';
import { useNewChatMenu } from '../../../../hooks/useNewChatMenu';
import { PopperMenu } from '../../../common';
import { MatterDialogState, useMatterActionsStore } from '../../../../store/useMatterActionsStore';
import useMFECommunication, { CustomEventData } from '../../../../hooks/useMFECommunication';

import './MatterSidebarHeader.scss';

interface MatterBodyHeaderProps {
  className?: string;
  isFolderView?: boolean;
  showAddChatButton?: boolean;
  onNewChat: () => void;
}

/**
 * Header component with new chat/folder and sorting capabilities.
 *
 * @param props - Component props.
 * @returns Rendered component.
 */
const MatterSidebarHeader = ({
  className,
  isFolderView,
  showAddChatButton = true,
  onNewChat
}: MatterBodyHeaderProps) => {
  const [newChatMenuState, newChatMenuToggle, newChatMenuRefElement, setNewChatMenuRef] = useNewChatMenu(false);
  const [
    sortMenuOptions,
    selectedSortOption,
    sortMenuState,
    sortMenuToggle,
    handelSortMenuItemClick,
    sortMenuRefElement,
    setMenuRefElement,
    setSortType,
    isLoading
  ] = useSortMenu(false);
  const [toggleMatterDialog] = useMatterActionsStore((state) => [state.toggleMatterDialog]);
  const [isOpenSort, setSortOpen] = useState(false);

  // Combine the two MFE communication listeners into one tp call sort left menu by date.
  const handleMFECommunication = (data: CustomEventData) => {
    if (data) {
       setSortType('by_date');
    }
  };

  useMFECommunication('isSortingMenu', handleMFECommunication);
  useMFECommunication('isChatting', handleMFECommunication);

  const sortMenuRef = useOutsideClick(() => {
    sortMenuState && sortMenuToggle();
  });
  const chatAndFolderMenuRef = useOutsideClick(() => {
    newChatMenuState && newChatMenuToggle();
  });

  return (
    <div className={classnames('matter-body-header', className)} data-testid="matter-body-header">
      {showAddChatButton && (
        <div
          className="new-folder-or-chat-button-container"
          ref={chatAndFolderMenuRef as RefObject<HTMLDivElement>}
          data-testid="new-folder-or-chat-button-container"
        >
          <Button
            appearance="primary"
            // onClick={isFolderView ? onNewChat : newChatMenuToggle}
            onClick={() => {
              onNewChat();
              newChatMenuToggle();
            }}
            data-testid="new-chat-button"
            // @ts-expect-error: The usePopper hook is designed to directly take DOM nodes instead of refs to update dynamically as nodes change. We use a callback ref and useState to support this dynamic behavior effectively.
            ref={setNewChatMenuRef}
          >
            <div className="new-folder-or-chat-button">
              <Icon iconName="plus" />
              <Translate tKey={isFolderView ? 'NEW_CHAT' : 'NEW_CHAT_OR_FOLDER_BUTTON'} />
            </div>
          </Button>
          {/* {newChatMenuState && !isFolderView && (
            <PopperMenu
              referenceElement={newChatMenuRefElement}
              className="new-folder-or-chat-menu"
              placement="bottom-start"
              strategy="fixed"
              dataTestId="new-folder-or-chat-menu"
            >
              <MenuItem
                data-testid="new-folder-or-chat-menu-item-chat"
                onClick={() => {
                  onNewChat();
                  newChatMenuToggle();
                }}
              >
                <div className="new-folder-or-chat-menu-item">
                  <Icon iconName="message-lines" />
                  <Translate tKey="NEW_CHAT" />
                </div>
              </MenuItem>
              <MenuItem
                onClick={() => {
                  toggleMatterDialog(MatterDialogState.NEW_MATTER, undefined);
                  newChatMenuToggle();
                }}
                data-testid="new-folder-or-chat-menu-item-folder"
              >
                <div className="new-folder-or-chat-menu-item">
                  <Icon iconName="folder" />
                  <Translate tKey="NEW_FOLDER" />
                </div>
              </MenuItem>
            </PopperMenu>
          )} */}
        </div>
      )}

      <div
        className="new-folder-or-chat-button-container"
        ref={sortMenuRef as RefObject<HTMLDivElement>}
        data-testid="sort-option-container"
      >
        <Button
           id='sort-option-btn'
           appearance="tertiary"
           onClick={() => {
              sortMenuToggle()
              setSortOpen(true)
            }}
           data-testid="sort-option-button"
           aria-expanded={isOpenSort ? 'true' : 'false'}
           aria-haspopup="true"
           aria-controls="sort-option-menu"
           disabled={isLoading}
           // Add a title to explain why the button is disabled
           title={isLoading ? 'Loading chat data...' : 'Sort chats'}
          // @ts-expect-error: The usePopper hook is designed to directly take DOM nodes instead of refs to update dynamically as nodes change. We use a callback ref and useState to support this dynamic behavior effectively.
          ref={setMenuRefElement}
        >
          <div className="new-folder-or-chat-button">
            <Translate tKey={sortMenuOptions.find((option) => option.value === selectedSortOption)?.title || ''} />
            {isLoading ? (
              /* Use a static spinner icon without animation */
              <Icon iconName="spinner" />
            ) : (
              <Icon iconName="angle-down" />
            )}
          </div>
        </Button>
        {sortMenuState && (
          <PopperMenu
            className="new-folder-or-chat-menu"
            dataTestId="sort-option-menu"
            referenceElement={sortMenuRefElement}
            placement="bottom-start"
            strategy="fixed"
          >
            {sortMenuOptions.map((menuOption) => (
              <MenuItem
                onClick={() => handelSortMenuItemClick(menuOption.value, setSortType)}
                data-testid={`sort-option-menu-${menuOption.value}`}
                key={`sort-option-menu-${menuOption.value}`}
                onKeyDown={(event) => {
                  if (event.key === 'Enter') {
                    handelSortMenuItemClick(menuOption.value, setSortType)
                  }
                  if (event.key === 'Tab' || event.key === 'Escape') {
                    sortMenuToggle();
                  }
                }}
              >
                <div className="new-folder-or-chat-menu-item">
                  <Translate tKey={menuOption.title} />
                </div>
              </MenuItem>
            ))}
          </PopperMenu>
        )}
      </div>
    </div>
  );
};

export default MatterSidebarHeader;
