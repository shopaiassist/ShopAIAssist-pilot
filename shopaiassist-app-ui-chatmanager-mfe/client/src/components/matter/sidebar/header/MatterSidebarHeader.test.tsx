import React from 'react';
import { renderWithI18n, snapshot, waitFor } from '../../../../../jest-utils';
import MatterSidebarHeader from './MatterSidebarHeader';
import { act } from 'react-dom/test-utils';
import axios from 'axios';
import { mockTree } from '../../../../utils/test-utils/mockData';

const getMockedComponent = (opts: { mockOnNewChat: () => void }) => {
  return <MatterSidebarHeader onNewChat={opts.mockOnNewChat} />;
};

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('MatterSidebarHeader Tests', () => {
  const mockOnNewChat = jest.fn();

  const sortOptionTests = async (sortOption: string) => {
    const component = getMockedComponent({ mockOnNewChat });
    const wrapper = renderWithI18n(component);
    const sortOptionsButton = await wrapper.getByTestId('sort-option-button');
    sortOptionsButton.click();
    await act(() => waitFor(() => wrapper.rerender(component)));
    const sortOptionSelect = await wrapper.getByTestId(`sort-option-menu-${sortOption}`);
    sortOptionSelect.click();
    await act(() => waitFor(() => wrapper.rerender(component)));
    expect(mockedAxios.get).toHaveBeenCalledWith('jest-test-domain/api/list', { params: { sortType: sortOption } });
    const sortMenu = await wrapper.queryAllByTestId('sort-option-menu');
    expect(sortMenu).toEqual([]);
  };

  describe('render', () => {
    it('should render', async () => {
      snapshot(getMockedComponent({ mockOnNewChat }));
    });
  });

  describe('user actions', () => {
    it('should open menu for new chat/folder', async () => {
      const component = getMockedComponent({ mockOnNewChat });
      const wrapper = renderWithI18n(component);
      const newChatButton = await wrapper.getByTestId('new-chat-button');
      newChatButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const newChatMenu = await wrapper.getByTestId('new-folder-or-chat-menu');
      expect(newChatMenu).toBeVisible();
    });

    it('should open menu for sort options', async () => {
      const component = getMockedComponent({ mockOnNewChat });
      const wrapper = renderWithI18n(component);
      const sortOptionsButton = await wrapper.getByTestId('sort-option-button');
      sortOptionsButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const sortMenu = await wrapper.getByTestId('sort-option-menu');
      expect(sortMenu).toBeVisible();
    });

    for (const sortOption of ['by_date', 'by_type', 'by_name']) {
      it(`should call function to change sort option (${sortOption})`, async () => {
        mockedAxios.get.mockResolvedValue(mockTree);
        await sortOptionTests(sortOption);
      });
    }
  });
});
