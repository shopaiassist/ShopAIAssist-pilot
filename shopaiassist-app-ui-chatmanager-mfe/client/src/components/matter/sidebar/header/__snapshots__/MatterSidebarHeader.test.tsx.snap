// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatterSidebarHeader Tests render should render 1`] = `
<div
  class="matter-body-header"
  data-testid="matter-body-header"
>
  <div
    class="new-folder-or-chat-button-container"
    data-testid="new-folder-or-chat-button-container"
  >
    <button
      appearance="primary"
      data-testid="new-chat-button"
    >
      <div
        class="new-folder-or-chat-button"
      >
        <div
          iconname="plus"
        />
        New
      </div>
    </button>
  </div>
  <div
    class="new-folder-or-chat-button-container"
    data-testid="sort-option-container"
  >
    <button
      appearance="tertiary"
      data-testid="sort-option-button"
    >
      <div
        class="new-folder-or-chat-button"
      >
        Sort by date
        <div
          iconname="angle-down"
        />
      </div>
    </button>
  </div>
</div>
`;
